### PR TITLE
Add ts support in parser

### LINKED ISSUE #345: Add support for typescript compilation on the fly for tests
Following where this appeared in #309 's [comment](https://github.com/paritytech/zombienet/issues/309#issuecomment-1231253582) by [@AuroraLantean[(https://github.com/AuroraLantean), we should add TS support for tests.

This could be implemented, based on extension and use `tsc` to build on the fly

### PR DESCRIPTION
**Pull request by wirednkod**
Fixes #345

### CODE REVIEW

File: javascript/packages/orchestrator/src/test-runner/assertions.ts [hunk 1]
```diff
@@ -232,7 +234,16 @@ const CustomJs = ({
         : custom_args.split(",")
       : [];
 
-    const resolvedJsFilePath = path.resolve(configBasePath, file_path!);
+    let resolvedJsFilePath = path.resolve(configBasePath, file_path!);
+
+    if (is_ts) {
+      await execa.command(`tsc ${resolvedJsFilePath}`);
```
Comment (pepoviola) [inline javascript/packages/orchestrator/src/test-runner/assertions.ts:240]: nice :)

Comment (wirednkod) [reply javascript/packages/orchestrator/src/test-runner/assertions.ts:240]: That makes sense. Ill alter accordingly

File: crates/parser/src/lib.rs [hunk 2]
```diff
@@ -16,6 +16,12 @@ use ast::{Assertion, AssertionKind, Comparison, NodeName, ParaId, TestDefinition
 #[cfg(test)]
 mod tests;
 
+enum ScriptType {
+    JS,
+    TS,
+    SH,
+}
```
Comment (l0r1s) [inline crates/parser/src/lib.rs:23]: How about being more explicit ?

```suggestion
enum ScriptType {
    Javascript,
    Typescript,
    Shell,
}
```

File: crates/parser/src/lib.rs [hunk 3]
```diff
@@ -155,7 +161,10 @@ fn parse_lines_count_match_pattern_rule(
     Ok((name, match_type, pattern, comparison, timeout))
 }
 
-fn parse_custom_script_rule(record: Pair<Rule>, is_js: bool) -> Result<AssertionKind, ParserError> {
+fn parse_custom_script_rule(
+    record: Pair<Rule>,
+    script_type: ScriptType,
+) -> Result<AssertionKind, ParserError> {
```
Comment (l0r1s) [inline crates/parser/src/lib.rs:167]: I think all these methods should be implemented on Pair<Rule>, for example:

```rust
impl Pair<Rule> {
     fn parse_custom_script_rule(&self, script_type: ScriptType) -> Result<AssertionKind, ParserError> {
       todo!()
     }
}
```

Then you can make the code cleaner:
```rust
let record = Pair {};

record.parse_custom_script_rule()?;
```

**Review by l0r1s — COMMENTED** [review]
Some improvements for readability possible but lgtm

File: crates/parser/src/lib.rs (hunk 3 above)
Comment (wirednkod) [reply crates/parser/src/lib.rs:167]: We can do that in a different PR

Comment (l0r1s) [reply crates/parser/src/lib.rs:167]: Maybe just add a #[TODO] ?

Comment (wirednkod) [reply crates/parser/src/lib.rs:167]: Even better :) I'll add an issue 😉

File: javascript/packages/orchestrator/src/test-runner/assertions.ts [hunk 4]
```diff
@@ -232,7 +236,23 @@ const CustomJs = ({
...
+        compilerOptions: { module: ts.ModuleKind.CommonJS },
+      });
+
+      resolvedJsFilePath = path.resolve(
+        configBasePath,
+        path.parse(file_path!).name + ".js",
+      );
+      await fs.writeFile(
+        resolvedJsFilePath,
+        new Uint8Array(Buffer.from(result.outputText)),
```
Comment (pepoviola) [inline javascript/packages/orchestrator/src/test-runner/assertions.ts:253]: isn't `result.outputText` a string? I think we can use it directly.

File: tests/0008-custom-scripts.toml [hunk 5]
```diff
@@ -11,11 +11,11 @@ chain = "rococo-local"
   [[relaychain.nodes]]
   name = "bob"
 
-[[parachains]]
-id = 100
-add_to_genesis = true
+# [[parachains]]
```
Comment (pepoviola) [inline tests/0008-custom-scripts.toml:14]: If this test doesn't require a para we should remove this lines.

**Review by pepoviola — COMMENTED** [review]
Looks good! Couple of questions, thanks!!

**Review by l0r1s — APPROVED** [review]

**Pull request merged**